// @uniswap/sdk/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var bignumber_js_1 = __importDefault(require("bignumber.js"));
exports.BigNumber = bignumber_js_1.default;
var constants_1 = require("./constants");
exports.ETH = constants_1.ETH;
exports.SUPPORTED_CHAIN_ID = constants_1.SUPPORTED_CHAIN_ID;
exports.FACTORY_ADDRESS = constants_1.FACTORY_ADDRESS;
exports.FACTORY_ABI = constants_1.FACTORY_ABI;
exports.EXCHANGE_ABI = constants_1.EXCHANGE_ABI;
exports.TRADE_TYPE = constants_1.TRADE_TYPE;
exports.TRADE_EXACT = constants_1.TRADE_EXACT;
exports.TRADE_METHODS = constants_1.TRADE_METHODS;
exports.TRADE_METHOD_IDS = constants_1.TRADE_METHOD_IDS;
exports.FIXED_UNDERFLOW_BEHAVIOR = constants_1.FIXED_UNDERFLOW_BEHAVIOR;
__export(require("./data"));
__export(require("./computation"));
__export(require("./format"));
__export(require("./orchestration"));
__export(require("./transact"));
